### PR TITLE
Ngram scorer update 

### PR DESCRIPTION
**Pull request by qmpzzpmq**
1.  glitch fix
2.  added an option for full or part ngram scorer for decoding acceleration
3. and transform it to local score
4. RESULT update, lower test WER from 6.7% to 6.6%

### CODE REVIEW

**Comment by qmpzzpmq** [comment]
@sw005320 
hi, shinji, could you give me feedback about this Codecov issue? I am not good at it. But I cleaned my commit log this time

**Comment by sw005320** [comment]
Cool!
Thanks for the update!

> @sw005320
> hi, shinji, could you give me feedback about this Codecov issue? I am not good at it. But I cleaned my commit log this time

You can ignore this.
This will be passed when you add some new classes or functions with appropriate documentation. These types of modifications are not important.

File: espnet/bin/asr_recog.py [hunk 1]
```diff
@@ -174,6 +174,13 @@ def get_parser():
         "--ngram-model", type=str, default=None, help="ngram model file to read"
     )
     parser.add_argument("--ngram-weight", type=float, default=0.1, help="ngram weight")
+    parser.add_argument(
+        "--ngram-scorer",
+        type=str,
+        default="full",
+        choices=("full", "part"),
+        help="ngram scorer choices",
```
Comment (sw005320) [inline espnet/bin/asr_recog.py:182]: Could you explain a bit more about the full and part scorer options?

File: egs/aishell/asr1/RESULTS.md [hunk 2]
```diff
@@ -40,17 +40,17 @@ exp/train_sp_pytorch_no_patience/decode_test_beam20_emodel.acc.best_p0.0_len0.0-
...
+|   Sum/Avg  |   14326      205341  |   94.1        5.7        0.2        0.1        6.0      41.7    |
+exp/train_sp_pytorch_train_pytorch_transformer_lr1.0/decode_test_decode_pytorch_transformer_lm0.7_4gram_0.3/result.txt
+|   SPKR     |   # Snt      # Wrd   |   Corr        Sub        Del        Ins        Err      S.Err   |
+|   Sum/Avg  |   7176       104765  |   93.5        6.3        0.2        0.1        6.6      44.6    |
 ```
   - only e2e model
 exp/train_sp_pytorch_train_pytorch_transformer_lr1.0/decode_dev_decode_pytorch_transformer/result.txt
 |   SPKR     |   # Snt      # Wrd   |   Corr        Sub        Del        Ins        Err      S.Err   |
 |   Sum/Avg  |   14326       205341 |   93.6        6.2        0.2        0.1        6.5      45.6    |
 exp/train_sp_pytorch_train_pytorch_transformer_lr1.0/decode_test_decode_pytorch_transformer/result.txt
```
Comment (sw005320) [inline egs/aishell/asr1/RESULTS.md:54]: These lines should be in the coding mode. Add "```", i.e., 
```
exp/train_sp_pytorch_train_pytorch_transformer_lr1.0/decode_dev_decode_pytorch_transformer/result.txt
|   SPKR     |   # Snt      # Wrd   |   Corr        Sub        Del        Ins        Err      S.Err   |
|   Sum/Avg  |   14326       205341 |   93.6        6.2        0.2        0.1        6.5      45.6    |
exp/train_sp_pytorch_train_pytorch_transformer_lr1.0/decode_test_decode_pytorch_transformer/result.txt
|   SPKR     |   # Snt      # Wrd   |   Corr        Sub        Del        Ins        Err      S.Err   |
|   Sum/Avg  |   7176       104765  |   92.7        7.1        0.2        0.1        7.4      49.8    |
```

**Comment by sw005320** [comment]
Sorry to ask many things.
Is it possible to add a partial scorer test as well?

File: espnet/bin/asr_recog.py (hunk 1 above)
Comment (qmpzzpmq) [reply espnet/bin/asr_recog.py:182]: If the ngram is set as a part scorer, similar to CTC scorer, the ngram scorer only score to topID.
but ngram is set as a full scorer, the ngram score all ID.

**Comment by qmpzzpmq** [comment]
@sw005320 I will try

File: espnet/bin/asr_recog.py (hunk 1 above)
Comment (sw005320) [reply espnet/bin/asr_recog.py:182]: Oh sorry, I meant to say that it’s better to add such descriptions. Probably in the help message?

Comment (qmpzzpmq) [reply espnet/bin/asr_recog.py:182]: good idea, I will add it.

**Comment by qmpzzpmq** [comment]
@sw005320 done
ngram full and part scorer have the same result but the full scorer takes 9860s in dev set decoding part one takes only 4090s in my test.

File: egs/aishell/asr1/RESULTS.md [hunk 3]
```diff
@@ -40,17 +40,27 @@ exp/train_sp_pytorch_no_patience/decode_test_beam20_emodel.acc.best_p0.0_len0.0-
 # Ngram related 
    - there is no RNN not ngram
```
Comment (sw005320) [inline egs/aishell/asr1/RESULTS.md:41]: Could you change this sentence more clearly?

File: espnet/bin/asr_recog.py [hunk 4]
```diff
@@ -174,6 +174,16 @@ def get_parser():
         "--ngram-model", type=str, default=None, help="ngram model file to read"
     )
     parser.add_argument("--ngram-weight", type=float, default=0.1, help="ngram weight")
+    parser.add_argument(
+        "--ngram-scorer",
+        type=str,
+        default="full",
```
Comment (sw005320) [inline espnet/bin/asr_recog.py:180]: How about making "part" as default?
I think most scenario uses "part".

**Comment by sw005320** [comment]
Many thanks!

**Pull request merged**